cron/jobstats.go

```diff
@@ -0,0 +1,91 @@
+package cron
+
+import (
+	"database/sql"
+)
+
+type JobStats struct {
+	Category	string
+	Status		string
+	Count		int
+}
+
+//jobs/counts
+func GetJobCounts(db *sql.DB, startDate string, endDate string)([]JobStats, error){
+
+	query := `WITH internal_app_ids AS (
```

Comment:
This is not something to work on now, but it's possible this query could be optimized, in particular if we can join the two halves of the UNION. Since they're filtering on different date fields, it might not be terribly advantageous though (in particular, it might force sequential scans rather than index scans, which could be bad for lots of data). Just leaving a quick note here for now so it can be looped back around to later if there's a need or there's interest.

With the current query, using this january 1 - 10 as the dates, https://explain.depesz.com/s/OA7UT is the plan on the QA DB, and https://explain.depesz.com/s/lP4l against production.

Reply:
Heh, looking at it, combining the two sides of the UNION is more complicated than I thought, so it might not be possible (I was misreading the query a little). It's still possibly something that can be improved, but the performance isn't bad anyway at present, so I suspect it won't be worth the effort until/unless we run into issues.

Reply:
I'm actually a little relieved you came to that conclusion. My initial impression was that a UNION would be the only sane way to do this, so that's what I used when I suggested combining the individual queries from the `de-stats` command-line utility. 😌

Reply:
Yeah -- I'd neglected to notice (or properly process) that while the first half of the union is for the 'Submitted' status, it doesn't actually filter down to it (because it needs to include jobs that are running or which are completed, and just filter by _when_ they were submitted). So since jobs can appear both in a submitted count and a terminal-status one, we'd have to do some crazy gymnastics to get the counts, I think. I'd originally misread it as the only difference being filtering on start_date vs. filtering on end_date, which is pretty simple to combine.

Reply:
Ah, that makes sense.